design/layer2-prefer-leader.md

```diff
@@ -0,0 +1,103 @@
+# Configuring Preferred Node To Be Layer2 Leader
+
+## Summary
+
+The purpose of the ehancement is to support configuring preferred node to be the leader for L2 mode.
+
+With this enhancement, the leader election will be artificially controllable.
+
+## Motivation
+
+MetalLB’s layer2 mode is simliar to `Keepalived`(https://metallb.universe.tf/concepts/layer2/#comparison-to-keepalived) which using `VRRP`.
+
+This enhancement implement functionality similar to [Priority and Preemption](https://www.cisco.com/assets/sol/sb/Switches_Emulators_v2_3_5_xx/help/350_550/index.html#page/tesla_350_550_olh/ts_vrrp_18_09.html) configuration of `VRRP`.
+
+### Goals
+
+- In L2 mode, being able to specify a node as the preferred leader for a service.
+- When a node is configured as the preferred leader, it replaces the current leader.
+
+### Non-Goals
+
+- Change the current leader mechanism election using memberlist.
+- Change the data models of Metallb's CRDs.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+When using Metallb's Layer2 mode, user wants to specify one node as the leader for a service and let all traffic for this service's IP goes to this sepcified node.
```

Comment:
This still doesn't tell anything about the specific use case. What I would like to understand is, is there a reason why a user would want to prefer a node over another but not strictly, that justifies the addition of the feature?

Reply:
We'd like to make MetalLB provide the similar function as the one provided by priority in VRRP, which gives user the feasibility to select the node they prefer to serve the traffic. Traditionally, this can be done by setting priorities. We consider the same requirement applies to VIP function provided by MetallB. The node preferred may have some special characters (resources/connectivity/service affinity/...) defined by user to make it the best candidate during an election. Hash-based election may not be able to reflect user's wish. 

We can also define a list of nodes to define the preferences of them. But so far a single one can also satisfy our needs. Not sure if someone would need it.

Reply:
This is what I am asking. Changing the code "to make it work like vrrp" is not a valid reason. 
We need to document and discuss what is the use case and what benefit will this change provide to metallb's users, and if that benefit is worth the extra complexity added to the codebase.